client/app/services/query.js

```diff
@@ -50,7 +50,7 @@ function isDateRangeParameter(paramType) {
 }
 
 export class Parameter {
-  constructor(parameter) {
+  constructor(parameter, parentQueryId) {
```

Comment:
Suggestion: add a new method to parameter, something like:

```javascript
loadDropdownValues() {
  if (this.ownerQueryId) {
    return Query.associatedDropdown({ queryId: this.ownerQueryId, dropdownQueryId: this.queryId }).$promise;
  } else {
    return Query.asDropdown({ id: this.queryId }).$promise;
  }
}
```

It will help to avoid passing `parentQueryId` (BTW - don't call it "parent" - the query rather "owner" of parameter) through component hierarchy - you can just use `parameter.loadDropdownValues().then(...)`. You'll still need to pass extra prop to `QueryBasedParameterInput`, but I think you can just replace `value` prop with entire parameter.

Reply:
Nice! How's 2be935d9?

Reply:
I think it's much better now :+1: